server/src/main/java/org/elasticsearch/common/blobstore/RetryingInputStream.java

```diff
@@ -309,11 +319,44 @@ protected interface BlobStoreServices {
      * Represents an {@link InputStream} for a single attempt to read a blob. Each retry
      * will attempt to create a new one of these. If reading from it fails, it should not retry.
      */
-    protected abstract static class SingleAttemptInputStream extends InputStream {
+    protected static final class SingleAttemptInputStream<V> extends FilterInputStream {
```

Comment:
I think the best approach for the `SingleAttemptInputStream` is to implement it as a decorator. This gives the CSPs more freedom in how they implement the single-attempt stream.

Specifically, if you want to do something on every read, it's much more work to extend `FilterInputStream` because the default implementations all delegate to the wrapped stream. If you extend `InputStream` you only need to implement `int read()` and the defaults are all implemented on top of that.

If we expect everyone to extend `SingleAttemptInputStream` we force everyone to extend whichever of the above that we extended. This is the inheritance issue I alluded to earlier due to `InputStream` being an abstract class rather than an interface.